### ISSUE TITLE
Multiple invalid warnings on some code

### ISSUE THREAD

**Issue by adriweb** [Benjamin-Dobell/IntelliJ-Luanalysis#31]
Environment
--------------
   name   |   version  
----------|----------
 IDEA version       | 2020.3 (IC-203.5981.155)
 Luanalysis version | 1.2.1
 OS                 | Linux Ubuntu 20.10

What are the steps to reproduce this issue?
------------------------------------------------
This code:
```lua
local utilities = {}

--- Split a string using pattern.
--- @overload fun(str:string):string[]
--- @param str string the string to split
--- @param pattern string pattern used to split. Default = "[^%s]+"
--- @return string[]
utilities.stringSplit = function(str, pattern)
  if type(str) ~= "string" then return {} end
  if type(pattern) ~= "string" or pattern:len() == 0 then
    pattern = "[^%s]+"
  end
  local parts = { __index = table.insert }
  setmetatable(parts, parts)
  str:gsub(pattern, parts)
  setmetatable(parts, nil)
  parts.__index = nil
  return parts
end
```

What happens?
-------------
1. When hovering `utilities.stringSplit`:
![image](https://user-images.githubusercontent.com/660896/103682710-38492980-4f57-11eb-8e60-4ee8046e25d8.png)
Note that this disappears if I annotate `utilities` with `---@type table` but I don't see why I'd need to do that though?

2. When hovering `parts.__index`:
![image](https://user-images.githubusercontent.com/660896/103682762-4ac36300-4f57-11eb-995f-4ff1230c7e92.png)
Why wouldn't I be able to set any field I want? Is it assuming a numeric table? It shouldn't considering it was created as a k-v table.

2. I was unable to repro this in a standalone lua file, but I've seen it as part of my utilities.lua file when hovering `parts` in the return:
![image](https://user-images.githubusercontent.com/660896/103682831-63cc1400-4f57-11eb-9b2a-aa2f2fe4c243.png)

What were you expecting to happen?
-----------------------------------------
No such warnings, it's all supposed to be valid as far as I know.

**Comment by Benjamin-Dobell** [comment]
I'm unable to reproduce this. It's possible there was an error somewhere and Luanalysis is in a bad state.

First, if you could for any error stacktraces that'd be great. Down the bottom right of IntelliJ there may a red exclamation mark icon (notifications). If there's something in you can share that'd be great.
 
Otherwise, can you reproduce this after `File` -> `Invalidate Caches / Restart...` -> `Invalidate and Restart`?

**Comment by Benjamin-Dobell** [comment]
Also, is this compiled from `master` or the official `1.2.1` release?

**Comment by adriweb** [comment]
Unfortunately, nothing in the event log / notifications, and same after invalidate&restart.
This is the official 1.2.1 release.

I doubt it helps more since you probably knew already, but the inspections' name are "Assign type" for point 2 and "Return type" for point 3.

**Comment by Benjamin-Dobell** [comment]
Think I found the bug.

I'm guessing this is in a file called `utilities.lua` and you have other files that are also called `utilities.lua`?

**Comment by adriweb** [comment]
Oh, that's true - this project in Intellij is a "meta-project" (not sure if that's the official name) and has all the other projects loaded too... Which makes all cross-dependencies work magically, but I guess creates this problem now.
Good catch !

What do you think a solution would be?

**Comment by Benjamin-Dobell** [comment]
Basically internally tables without an explicit type are given names like `<offset>@(<filename>)table`.

This is behaviour that originated in EmmyLua though, and it seems the file name is used in other circumstances too. So I'll need to properly go through and understand the repercussions before fixing it. It's definitely a bug that needs to be corrected in Luanalysis though.

For now you can work around it one of two ways:

1. Provide an explicit class name e.g. `---@class StringUtilities`.
2. Or, if you _really_ don't want to do that and are up for a _very_ hacky solution. You can ensure your two `local utilities = {}` declarations don't share the same offset in their file (comments, whitespace etc.)

**Comment by adriweb** [comment]
OK, I see, makes sense. Good luck for the proper fix then.
Thanks for the tips!

**Comment by adriweb** [comment]
Is that also related to the following issue somehow?

In the plugin settings, "Lua Additional Sources root" has project "Common"
```lua
-- in Project A, in foobar.lua:
local Constants = require("Common.Constants")
print(Constants.FOO)

-- in Project Common, in Constants.lua:
return {
  FOO = true,
  BAR = 42
}

```
OK: Ctrl-Clicking on the `Common.Constants` inside the require works fine and brings me to the out-of-project "Constants.lua" file.
FAIL: there's squiggly lines in the `Constants.FOO` print usage, and the error is: `No such member 'FOO' found on type 'table'`. Similarly, no autocompletion for `Constants.***`.

Everything is OK again if I just add `---@type table` at the top of Constants.lua.

I should note that this is new behaviour, before I updated today, there were no such "issues".

**Comment by adriweb** [comment]
In fact, it's also probably linked to all kinds of errors I'm getting in this meta-projects project.

A tiny repro:
```lua
local function foobar()
  local test = { 1, 2, 3 }
  test[2] = 0
end
```

In the meta-project intellij project:
![image](https://user-images.githubusercontent.com/660896/103697922-99c8c280-4f6e-11eb-997b-3aa30f49cd45.png)

In a standard intellij project, no such error anymore, although something else (that may be expected, but a bit strange?):
![image](https://user-images.githubusercontent.com/660896/103697887-8cabd380-4f6e-11eb-98ff-ab3c581e9143.png)

**Comment by Benjamin-Dobell** [comment]
Should be fixed in v1.2.2. Currently being reviewed by Jetbrains, but you can grab it from Github in the meantime.
